[FLINK-2901] Remove Record API dependencies from flink-tests

zentol opened the pull request:
This PR removes all dependencies to the Record API in the flink-tests project.

For this, tests were either ported or removed.

The following files were NOT ported:
* recordJobs/*
 * except kmeans/udfs CoordVector, PointInFormat (moved & ported)
 * except util/InfiniteIntegerInputFormat (moved & ported)
* recordJobTests/*
* operators/*
 *  except CoGroupSortITCase, MapPartitionITCase, ObjectReuseITCase (moved)
* accumulators/AccumulatorIterativeITCase
 * Unfinished test, should be addressed in a separate issue
* IterationTerminationWithTwoTails
 * nigh identical with IterationTerminationWithTerminationTail (ported version also failed)
* iterative/DeltaPageRankITCase, IterativeKMeansITCase, KMeansITCase
 * behaved like they belong into /recordJobTests/ (as in they simply execute a recordJob), thus removed
* optimizer/iterations/IterativeKMeansITCase
 * overlaps with ConnectedComponentsITCase (need a second opinion on this!)
* util/FailingTestBase
 * integrated into TaskFailureITCase (only user of the class)
* testPrograms/util/tests/*

cancelling/MatchJoinCancellingITCase was ported but disabled since it is unstable, and also disabled before. Separate issue seems appropriate.

fhueske commented:
Nice @zentol! Thanks for this effort!

fhueske commented on flink-tests/src/test/java/org/apache/flink/test/iterative/IterationTerminationWithTwoTails.java:
```diff
@@ -1,134 +0,0 @@
-import org.junit.Assert;
-
-@SuppressWarnings("deprecation")
-public class IterationTerminationWithTwoTails extends RecordAPITestBase {
```
I think we should keep this test and port it.

fhueske commented on flink-tests/src/test/java/org/apache/flink/test/iterative/IterationTerminationWithTwoTails.java:
```diff
@@ -1,134 +0,0 @@
-		iteration.setNextPartialSolution(sumReduce);
-		
-		MapOperator terminationMapper = MapOperator.builder(new TerminationMapper())
-				.input(iteration.getPartialSolution())
```
The difference to `IterationTerminationWithTerminationTail` is that the input of the `terminationMapper` is the partial solution and not the result of the `SumReducer`, right?

zentol replied:
so THAT was the difference, will port it!

fhueske commented on flink-tests/src/test/java/org/apache/flink/test/iterative/IterationWithAllReducerITCase.java:
```diff
@@ -18,91 +18,34 @@
-				.build();
-		
-		iteration.setNextPartialSolution(sumReduce);
+		DataSet<String> sumReduce = iteration.reduce(new ReduceFunction<String>(){
```
A `GroupReduceFunction` would be closer to the original test.

fhueske commented on flink-tests/src/test/java/org/apache/flink/test/iterative/IterativeKMeansITCase.java:
```diff
@@ -1,62 +0,0 @@
-import org.apache.flink.test.util.RecordAPITestBase;
-
-
-public class IterativeKMeansITCase extends RecordAPITestBase {
```
We should check if there is another test in the test suite that uses a broadcast variable in a bulk iteration. 
If not, we should implement such a test (not necessarily reimplementing the KMeans program).

zentol replied:
does the KMeansForTest qualify for this?

fhueske commented:
Hi @zentol, thanks again for the huge effort to port all these tests!

Looks mostly good, except for some minor things and three tests which should be ported to preserve test coverage, IMO.

zentol commented:
@fhueske I've addressed most of your concerns.

Things that still need work / clarification:
* PreviewPlanDumpTest was previously executed with 2 different sets of arguments, now only with 1. Should this be changed back to the previous behaviour? The arguments affect paths for sources/sink, parallelism and the number of Iterations
* test.classloading.jar.KMeansForTest appears to be a good replacement for the IterativeKMeansITCase, what's your take on that?
* The removed delta ilteration PageRank program looks very similar to the ConnectedComponents implementation under flink-examples. I don't think this needs a separate port.

fhueske commented:
Hi @zentol, sorry for the delayed review.
You are right, `DeltaPageRankITCase` (which executes for some reason the Record API delta iteration ConnectedComponents example) is identical to the DataSet `ConnectedComponentsITCase`. So we do not port this one.
I also agree to remove the checks for empty arguments in the `PreviewPlanDumpTest`. This requires cooperation of the job, i.e., the job may not fail for empty arguments, and tests exactly the same code paths as the tests with arguments, as far as I see.

However, I do not think that the `ClassLoaderITCase` which calls the `KMeansForTest` is a good replacement for the `IterativeKMeansITCase`. The programs are identical but, the `ClassLoaderITCase` only checks if the program executes without an error and does not check the result.

@zentol Is it OK for you if I add this one last test and merge this PR afterwards?

zentol commented:
Go ahead, i won't have time for it today anyway :)

fhueske commented:
Thanks, I'll port the last test and merge this PR
